### PR TITLE
Bash scripting for posix

### PR DESCRIPTION
**Pull request by julianoes**
## tl;dr

This adds bash scripting for POSIX. The startup file is now evaluated by bash.
Additionally, PX4 commands can be sent from other terminals, allowing commands such as `px4-commander status` or `px4-listener sensor_combined` from anywhere on the machine.
## Description

This PR adds the possibility to spawn PX4 modules out of bash. Basically,
the main executable can now be started as a server/daemon or as a
client.
The server replaces the existing functionality of the main exe with
the pxh shell, however, it also opens a pipe that clients can talk to.

Clients can run or spawn PX4 modules or commands by connecting to the
server over the pipe. They clients will get the stdout and return value
of their commands via a client specific pipe back.

The main exe starts as a server and runs a startup script using bash. Each command in the startup script such as 'commander start' gets aliased with a prefix as 'px4-commander start' and symlinked to the px4 main executable.
If the main executable is called with the prefix 'px4-' which is passed in `argv[0]`, it will act as a client and therefore send the command 'commander start' to the server. Once this command has been run in the server, the client will receive the return value and the bash script will continue with the next command.

Additionally, this allows commands to be sent from outside pxh, so from other terminals (on the same computer). All that is needed is to have the px4 exe and the symlinks in the PATH.

Thanks to @bkueng for the big help.

Let me know what you think: @jgoppert, @LorenzMeier, @dagar, @mcharleb 

Adresses: #4800
## Progress to get this PR in
- [x] Main SITL startup transferred to unified rcS startup script
- [ ] Additional SITL startups into rcS
- [ ] Check CI on jenkins
- [x] Script Snapdragon startup
- [ ] Script Bebop startup
- [ ] Script Raspberry Pi startup
- [ ] Upload mixers
- [ ] Stress test whole system with some scripts
- [x] Allow listener to keep looping
- [ ] Add Devguide documentation
## Future work after this PR
- Unify NuttX and Posix startup
### SITL

The normal SITL startup already uses the new unified SITL startup script rcS, it can get started with:

```
make posix gazebo
```

```
make posix jmavsim
```

To send a command in another terminal window, adapt the PATH:

```
export PATH=$PATH:~/src/Firmware/build_posix_sitl_default/bin
```

and send a command:

```
px4-commander status
```

Or listen to a uORB message:

```
px4-listener vehicle_attitude
```

Press Ctrl+C to exit.
### Snapdragon

To test this on Snappy, just build and upload normally:

```
make eagle_default upload
```

Open adb shell, and start it with:

```
cd /home/linaro/bin
export PATH=$PATH:/home/linaro/bin
./px4 rcS
```

Then choose, the airframe config using the param SYS_AUTOSTART and restart. The current options are:

```
SYS_AUTOSTART 4100: generic (flight) Quad
SYS_AUTOSTART 4200: Microheli 200QX
```

### CODE REVIEW

**Comment by dagar** [comment]
Very cool. For startup scripts I wonder if there's a subset of bash and nsh we could safely use and then completely unify startup.

**Comment by julianoes** [comment]
> For startup scripts I wonder if there's a subset of bash and nsh we could safely use and then completely unify startup.

Yer that would be the plan. Most commands of the nsh are already in bash, however, I found `set` which is "missing".

**Comment by bharath374** [comment]
Julian, this is a great solution!

One comment: The following lines should stay under "if param compare SYS_AUTOSTART 4100" since they're not been tested in other configurations.

qshell gps start -d /dev/tty-4
qshell df_hmc5883_wrapper start
qshell df_trone_wrapper start

File: cmake/CMakeLists_posix.txt [hunk 1]
```diff
@@ -160,6 +177,34 @@ endforeach()
 
 install(TARGETS px4 DESTINATION bin)
 install(DIRECTORY ${PROJECT_SOURCE_DIR}/ROMFS DESTINATION ${CMAKE_INSTALL_DATADIR}/${PROJECT_NAME})
-install(DIRECTORY ${PROJECT_SOURCE_DIR}/posix-configs DESTINATION ${CMAKE_INSTALL_DATADIR}/${PROJECT_NAME})
+install(DIRECTORY ${PROJECT_SOURCE_DIR}/posix-configs/SITL/init DESTINATION ${CMAKE_BINARY_DIR}/rootfs/etc/init)
```
Comment (bkueng) [inline cmake/CMakeLists_posix.txt]: Is this correct? CMAKE_BINARY_DIR seems wrong

File: cmake/CMakeLists_posix.txt [hunk 2]
```diff
@@ -160,6 +177,34 @@ endforeach()
...
+	)
+
+add_custom_command(TARGET px4
+	POST_BUILD
+	COMMAND mkdir -p ${CMAKE_BINARY_DIR}/rootfs/etc/
+	)
+
+add_custom_command(TARGET px4
+	POST_BUILD
+	COMMAND cp -R ${PROJECT_SOURCE_DIR}/posix-configs/SITL/init ${CMAKE_BINARY_DIR}/rootfs/etc/
```
Comment (bkueng) [inline cmake/CMakeLists_posix.txt]: We shouldn't use the POST_BUILD for these. It does not handle dependencies & changes to the script files. We should explicitly enumerate these files and use the `add_custom_command(OUTPUT` command, similar to the .msg files.

Then instead of `cp`, better would be `COMMAND ${CMAKE_COMMAND} -E copy`

File: Tools/adb_upload.sh [hunk 3]
```diff
@@ -5,13 +5,17 @@ if [[ "$#" < 2 ]]; then
...
+for last; do true; done
+
 echo "Wait for device..."
 adb wait-for-device
-echo "Uploading..."
 
-# Get last argument
-for last; do true; done
+echo "Creating folder structure..."
+#echo "creating folder: $last"
```
Comment (bkueng) [inline Tools/adb_upload.sh]: debug leftover?

File: cmake/posix/px4_impl_posix.cmake [hunk 4]
```diff
@@ -105,6 +105,64 @@ function(px4_posix_generate_builtin_commands)
...
+				set(${property} ${${property}_DEFAULT})
+			endif()
+		endforeach()
+		if (MAIN)
+			set(ln_name "${PREFIX}${MAIN}")
+			add_custom_command(TARGET ${TARGET}
+				POST_BUILD
+				COMMAND ln -s -f ${TARGET} ${ln_name}
+				WORKING_DIRECTORY "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}"
+				USES_TERMINAL
```
Comment (bkueng) [inline cmake/posix/px4_impl_posix.cmake]: I did not find `USES_TERMINAL`. Is it some magic undocumented variable?

File: src/platforms/posix/main.cpp [hunk 5]
```diff
@@ -32,134 +32,236 @@
...
-	string command = appargs[0];
+		if (!is_already_running()) {
+			PX4_ERR("PX4 daemon not running yet");
+			return -1;
+		}
 
-	if (apps.find(command) != apps.end()) {
-		const char *arg[appargs.size() + 2];
+		/* Remove the prefix by moving argv[0] and 1 more for the 0 termination. */
+		memmove(argv[0], argv[0] + strlen(prefix), strlen(argv[0]) + 1);
```
Comment (bkueng) [inline src/platforms/posix/main.cpp]: how about `argv[0] += strlen(prefix)` ?

File: src/platforms/posix/main.cpp [hunk 6]
```diff
@@ -32,134 +32,236 @@
...
+
+		register_sig_handler();
+		set_cpu_scaling();
+
+		DriverFramework::Framework::initialize();
+
+		px4::init_once();
+		px4::init(argc, argv, "px4");
+
+		run_startup_bash_script(commands_file);
```
Comment (bkueng) [inline src/platforms/posix/main.cpp]: This does not work with chroot for 2 reasons:
- bash will not be found when chrooting
- the commands file is opened for test before chrooting and then afterwards again. one of them will probably fail, since chroot changes the root.

do we even need chroot?

File: src/platforms/posix/px4_daemon/server.cpp [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+	}
+}
+
+void
+Server::_cleanup_thread(const uint64_t client_uuid)
+{
+	pthread_t pthread_killed = _client_uuid_to_pthread[client_uuid];
+	int pipe_fd = _pthread_to_pipe_fd[pthread_killed];
+
+	close(pipe_fd);
```
Comment (bkueng) [inline src/platforms/posix/px4_daemon/server.cpp]: should we `unlink` the pipe as well here? If I see correctly a client pipe is never deleted.

File: src/platforms/posix/px4_daemon/client.cpp [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+Client::_prepare_recv_pipe()
+{
+	int ret = get_client_recv_pipe_path(_uuid, _recv_pipe_path, sizeof(_recv_pipe_path));
+
+	if (ret < 0) {
+		PX4_ERR("failed to assemble path");
+		return ret;
+	}
+
+	unlink(_recv_pipe_path);
```
Comment (bkueng) [inline src/platforms/posix/px4_daemon/client.cpp]: I think this is a precaution in case a client with the same uuid exists/ed? I think we have to be careful here (however unlikely the case is): will this kill the other client? If so, we might be better off trying a different uuid.

File: src/platforms/posix/px4_daemon/history.cpp [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+	if (line.empty()) {
+		return;
+	}
+
+	// Don't add duplicate entries.
+	if (!_history.empty() && line.compare(_history.back()) == 0) {
+		return;
+	}
+
+	_history.push_back(line);
```
Comment (bkueng) [inline src/platforms/posix/px4_daemon/history.cpp]: use a maximum history length?

File: src/platforms/posix/px4_daemon/server.cpp [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+	}
+
+	// Run the actual command.
+	int retval = Pxh::process_line(message_str, false);
+
+	// Report return value.
+	_send_retval(pipe_fd, retval, client_uuid);
+
+	// Clean up before returning.
+	_instance->_cleanup_thread(client_uuid);
```
Comment (bkueng) [inline src/platforms/posix/px4_daemon/server.cpp]: This is not thread-safe as it accesses the maps from another thread.

File: src/platforms/posix/px4_daemon/server_io.cpp [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+	client_recv_packet_s *packet = &thread_data_ptr->packet;
+	packet->header.payload_length = buffer_length;
+
+	int pipe_fd = thread_data_ptr->pipe_fd;
+	int bytes_to_send = get_client_recv_packet_length(packet);
+
+	// Check if we can write first by writing 0 bytes.
+	// If we don't do this, we'll get SIGPIPE and be very unhappy
+	// because the whole process will go down.
+	int ret = write(pipe_fd, NULL, 0);
```
Comment (bkueng) [inline src/platforms/posix/px4_daemon/server_io.cpp]: Is this still needed, even if we ignore `SIGPIPE`?

**Comment by bkueng** [comment]
Finished reviewing, added some small obvious changes. Comments adress mostly small stuff, I think the cmake needs some more work. But overall really good code!

File: src/platforms/posix/main.cpp (hunk 6 above)
Comment (julianoes) [inline src/platforms/posix/main.cpp]: Also, you don't have permission to chroot without being root, right?

I don't think we need chroot, and @LorenzMeier doesn't think either.

File: src/platforms/posix/px4_daemon/server.cpp (hunk 7 above)
Comment (julianoes) [inline src/platforms/posix/px4_daemon/server.cpp]: We just unlink before trying to create it, is that not enough?

File: src/platforms/posix/px4_daemon/client.cpp (hunk 8 above)
Comment (julianoes) [inline src/platforms/posix/px4_daemon/client.cpp]: Good point, I'd say just fail in the case it exists already.

File: src/platforms/posix/px4_daemon/server.cpp (hunk 7 above)
Comment (julianoes) [inline src/platforms/posix/px4_daemon/server.cpp]: Ah, that's the specific one, yer let's delete it.

File: src/platforms/posix/px4_daemon/history.cpp (hunk 9 above)
Comment (julianoes) [inline src/platforms/posix/px4_daemon/history.cpp]: Why, to limit RAM usage?

File: src/platforms/posix/px4_daemon/server.cpp (hunk 10 above)
Comment (julianoes) [inline src/platforms/posix/px4_daemon/server.cpp]: So I need to add locking?

File: src/platforms/posix/px4_daemon/server_io.cpp (hunk 11 above)
Comment (julianoes) [inline src/platforms/posix/px4_daemon/server_io.cpp]: Yes, ignoring `SIGPIPE` did not have the effect we wanted and this is still necessary.

**Comment by julianoes** [comment]
@bkueng Thanks for the review!

File: cmake/posix/px4_impl_posix.cmake (hunk 4 above)
Comment (dagar) [inline cmake/posix/px4_impl_posix.cmake]: It's relatively new and only relevant for ninja. Without it ninja buffers all output only showing it once the command has finished.

Comment (bkueng) [inline cmake/posix/px4_impl_posix.cmake]: ok, thanks. But then it only makes sense for commands with lots of expected output.

File: src/platforms/posix/main.cpp (hunk 6 above)
Comment (bkueng) [inline src/platforms/posix/main.cpp]: Then we can remove it altogether. Except if we have a use-case for the install...

File: src/platforms/posix/px4_daemon/client.cpp (hunk 8 above)
Comment (bkueng) [inline src/platforms/posix/px4_daemon/client.cpp]: ok

File: src/platforms/posix/px4_daemon/history.cpp (hunk 9 above)
Comment (bkueng) [inline src/platforms/posix/px4_daemon/history.cpp]: yes. if there is some automated process feeding in commands, we don't want the history to grow infinitely.

File: src/platforms/posix/px4_daemon/server.cpp (hunk 10 above)
Comment (bkueng) [inline src/platforms/posix/px4_daemon/server.cpp]: yes, that's probably the simplest way. or signal the server thread to remove the uuid, but this also requires a queue with synchronization.

Comment (bkueng) [inline src/platforms/posix/px4_daemon/server.cpp]: yes, that's probably the simplest way. or signal the server thread to remove the uuid, but this also requires a queue with synchronization.

File: cmake/CMakeLists_posix.txt (hunk 1 above)
Comment (julianoes) [inline cmake/CMakeLists_posix.txt]: I'm gonna remove the install stuff again, not sure how this should work.

Comment (jgoppert) [inline cmake/CMakeLists_posix.txt]: @julianoes  I think it would be worthwhile to rebase on https://github.com/PX4/Firmware/pull/5255 now.

File: Tools/adb_upload.sh (hunk 3 above)
Comment (julianoes) [inline Tools/adb_upload.sh]: Thx, fixed.

File: cmake/posix/px4_impl_posix.cmake (hunk 4 above)
Comment (julianoes) [inline cmake/posix/px4_impl_posix.cmake]: Removed it since it's not really needed.

File: src/platforms/posix/main.cpp (hunk 5 above)
Comment (julianoes) [inline src/platforms/posix/main.cpp]: Aha, of course, that's easier. :facepunch:

File: src/platforms/posix/main.cpp (hunk 6 above)
Comment (julianoes) [inline src/platforms/posix/main.cpp]: Removed.

File: src/platforms/posix/px4_daemon/history.cpp (hunk 9 above)
Comment (julianoes) [inline src/platforms/posix/px4_daemon/history.cpp]: Ok, will do.

**Comment by lucasdemarchi** [comment]
I think I'm probably late for this, but I was working on the current state of master branch and then I saw this PR.  Quick question:  why do you want to start the px4 components via arbitrary shell scripts?  Wouldn't a declarative approach via configuration file (ini, json or other format) and then probably a single binary to control the internal parts be simpler?

File: src/platforms/posix/px4_daemon/server.cpp (hunk 10 above)
Comment (julianoes) [inline src/platforms/posix/px4_daemon/server.cpp]: Done.

**Comment by LorenzMeier** [comment]
Because it allows you to follow the UNIX system design of a collection of small apps and mixing calls to the system with calls to the PX4 middleware.

It's easy to support a JSON file format as well, but that would be rather for fixed vehicle configurations (which is important for products).

I suggest you try this PR when it's done - running the uORB listener in a separate she'll and watching different topics in parallel is an extremely powerful development tool.

**Comment by lucasdemarchi** [comment]
> Because it allows you to follow the UNIX system design of a collection of small apps and mixing calls to the system with calls to the PX4 middleware.

It's certainly an improvement over the current one.  However since the wrappers are actually talking to another daemon (the px4 binary) that responds accordingly, it's more a client/server approach than a classical "collection of small apps". In this case I usually see people using a declarative approach, i.e. a config file, because it avoids tempting hacks to add lots of sleeps here and there, scripting all the startup sequence. It's more or less a similar scenario of init implementation in which everybody went away from pure script init (be it systemd, upstart, android init, MacOS, etc) to a more declarative approach, even if in some of them still exists limited support for scripting.

**Comment by LorenzMeier** [comment]
I'm not at all dismissing the usefulness of a JSON based config. I guess we're really talking about different use cases for startup and tools / development. A JSON config would be extremely helpful.

**Comment by julianoes** [comment]
@jgoppert ok I finally got around to do a rebase. I have a few questions and points to discuss.
I think it would be helpful for me to sync in a call if you have time.

**Comment by LorenzMeier** [comment]
Nice! How far off do you feel is this?

**Comment by julianoes** [comment]
> Nice! How far off do you feel is this?

Have a look at the todos in the description, there is still quite a bit of work.

Still some work to be done. One big factor is that tt depends on how much of the startup scripts we want to merge already. I guess we can always do that in later steps.

**Comment by LorenzMeier** [comment]
I think its ok to not merge too much initially but just bring in the capability. Then we can refactor individual platforms piece by piece.

**Comment by julianoes** [comment]
Agreed. First up, I need to sync with @jgoppert because I'm pretty sure I broke all his work with the rebase.
Also, Pixhawk builds overflow flash space, not sure why.

**Comment by julianoes** [comment]
px4fmu-v2_default build:
```
master: 1022856 bytes
feature_shell: 1023904 bytes
```
with
```
arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 5.4.1 20160919 (release) [ARM/embedded-5-branch revision 240496]
```

**Comment by jgoppert** [comment]
@julianoes, do you have time after the meeting tomorrow or right now to sync up? Is skype best?

**Comment by jgoppert** [comment]
One  thing to keep in mind is that it would be nice to specify the rootfs as ./vehicle1/rootfs ./vehicle2/rootfs. I was doing this and realized I was using the same params for both vehicles which made things interesting.

**Comment by julianoes** [comment]
@AndreasAntener I could use your help here. I have the unit tests working again (I think) but I'm struggling with the ROS integration tests. It builds but it fails to find the px4 node. I changed the path of the px4 binary from `build_posix_sitl_default/src/firmware/posix/px4` to `build_posix_sitl_default/bin/px4` so that might be why.

The error I'm getting is:
```
[roslaunch][ERROR] 2016-10-23 14:56:24,320: can't locate node [px4] in package [px4]
[rostest][ERROR] 2016-10-23 14:56:24,320: ERROR: cannot launch node of type [px4/px4]: can't locate node [px4] in package [px4]
```

The last changes trying to make this work are in https://github.com/PX4/Firmware/pull/5162/commits/f39a2f7e3f3042de6397145d8a6c260470512746.

I'm aware of the fact that the SITL stuff for VTOL is not done yet and won't work. I was trying to get iris flying first.

**Comment by AndreasAntener** [comment]
@julianoes on my and (and Jenkins) it looks more like it cannot even compile:
```
CMake Error at px4/cmake/common/px4_base.cmake:106 (message):
  px4_posix_generate_alias requires argument OUT
```

Maybe you have a stale state?

**Comment by julianoes** [comment]
@AndreasAntener strange, it does compile for me doing `integrationtests/run_tests.bash`. I deleted px4 in catkin/src/ but it didn't reproduce the jenkins cmake error either.

**Comment by mhkabir** [comment]
@julianoes Do we have any plans to get this in soon?

**Comment by julianoes** [comment]
@mhkabir Soonish, yes.

**Comment by julianoes** [comment]
Ok, I've rebased this once again. I was able to do `make posix jmavsim` and get it to takeoff. Most of the rest is presumably broken (again) though.

**Comment by dagar** [comment]
@julianoes I'm really looking forward to this. What's needed to get it in? The PR is massive, are there smaller pieces we could start breaking off and pulling in?

**Comment by julianoes** [comment]
@dagar: no, I think we need to consume it in one piece. Next steps is porting all the startup scripts into one. Let me plan this for this weekend.

**Comment by TSC21** [comment]
@julianoes do you have the required bandwidth to proceed with this. This would be a very important update for our interface with the flight controller.

**Comment by LorenzMeier** [comment]
@julianoes I'm closing this as stale. I would love to see this in by all means. But I think it would be good if you could break it down instead into one PR that adds the capability (but leaves the current startup scripts) and then iterate on breaking those down.

That should be much quicker and take much less of your time than re-architecting the complete startup of all targets.

Thanks!

**Comment by julianoes** [comment]
Agreed. I need a week of holiday for this one.
